src/Compatibility/Microsoft.DotNet.PackageValidation/ApiCompatRunner.cs

```diff
@@ -31,63 +36,88 @@ public ApiCompatRunner(string noWarn, (string, string)[] ignoredDifferences, boo
         }
 
         /// <summary>
-        /// Runs the api compat for the tuples in the queue.
+        /// Runs the api compat for the tuples in the dictionary.
         /// </summary>
         public void RunApiCompat()
         {
-            foreach (var apicompatTuples in _queue.Distinct())
+            foreach (MetadataInformation left in _dict.Keys)
             {
-                // TODO: Add optimisations tuples.
-                using (Stream leftAssemblyStream = GetFileStreamFromPackage(apicompatTuples.leftAssemblyPackagePath, apicompatTuples.leftAssemblyRelativePath))
-                using (Stream rightAssemblyStream = GetFileStreamFromPackage(apicompatTuples.rightAssemblyPackagePath, apicompatTuples.rightAssemblyRelativePath))
+                Stream leftAssemblyStream = GetFileStreamFromPackage(_leftPackagePath, left.AssemblyId);
+                IAssemblySymbol leftSymbols = new AssemblySymbolLoader().LoadAssembly(left.AssemblyName, leftAssemblyStream);
+                leftAssemblyStream.Close();
+                ElementContainer<IAssemblySymbol> leftContainer = new(leftSymbols, left); ;
+
+                List<ElementContainer<IAssemblySymbol>> rightContainerList = new();
+                foreach (var rightTuple in _dict[left])
                 {
-                    IAssemblySymbol leftSymbols = new AssemblySymbolLoader().LoadAssembly(apicompatTuples.assemblyName, leftAssemblyStream);
-                    IAssemblySymbol rightSymbols = new AssemblySymbolLoader().LoadAssembly(apicompatTuples.assemblyName, rightAssemblyStream);
+                    Stream rightAssemblyStream = GetFileStreamFromPackage(_rightPackagePath, rightTuple.rightAssembly.AssemblyId);
+                    IAssemblySymbol rightSymbols = new AssemblySymbolLoader().LoadAssembly(rightTuple.rightAssembly.AssemblyName, rightAssemblyStream);
+                    rightAssemblyStream.Close();
+                    rightContainerList.Add(new ElementContainer<IAssemblySymbol>(rightSymbols, rightTuple.rightAssembly));
+                }
 
-                    _log.LogMessage(MessageImportance.Low, apicompatTuples.header);
+                IList<(MetadataInformation, MetadataInformation, IEnumerable<CompatDifference>)> differences =
+                    _differ.GetDifferences(leftContainer, rightContainerList).ToList();
```

Comment:
Do we really need to convert this to a List? or can we just enumerate the result and have a counter to access the dictionary?